dotnet/corefx
# Cleanup ImmutableArray tests

This is still WIP, but since it's a rather large PR and almost finished I thought I'd submit it. There are still a few TODOs floating around, but feel free to review.

**Changes:**

- Theorize most of the tests to make it easier to add new test cases
- Segregate tests that assert equality from tests that check exceptions
- Standardize method/variable naming to make tests easier to read
- Fill in some gaps in the tests
  - Test more explicit interface implementations
  - Run methods that accept comparers over a variety of comparers
  - Run methods that accept collections over a variety of collections (arrays, lists, `IList<T>`, lazy enumerables, etc.)
- General cleanup
  - Replace `Assert.Equal(immutableArray1, immutableArray2)` with `Assert.True(immutableArray1 == immutableArray2)`
    - This has a poorer stack trace, but it makes it more explicit that we're testing reference equality

**Note:** I tried to keep many of the existing test cases we have, so that's why you'll see a lot of (sometimes redundant) cases in the MemberData.

/cc @AArnott @ianhays @stephentoub
